### PR TITLE
pat: add prefix pattern

### PR DESCRIPTION
**Pull request by timbray**
fixes: #62
Signed-off-by: Tim Bray <tbray@textuality.com>

### CODE REVIEW

**Comment by timbray** [comment]
Hmm, I guess I should bump version to 1.1.0, this is a new feature.

**Comment by timbray** [comment]
> is identical to
> 
> ```
> {"a": [ { "shellstyle": "foo*"} ] } 
> ```

Actually, I'm an idiot squared. That breaks if there's a `*` in the prefix. OK, am happy with this now, un-draft-i-fying it, would appreciate a review

File: README.md [hunk 1]
```diff
@@ -113,6 +113,14 @@ The following Patterns would match it:
   }
 }
 ```
+```json
+{
+  "Image": {
+    "Thumbnail": {
+      "Url": [ "a", { "prefix": "https:" } ] }
```
Comment (embano1) [inline README.md:120]: nit: 
```suggestion
      "Url": [ "a", { "prefix": "http:" } ] }
```

I guess you did that one on purpose to see if I'm reviewing :)

File: pattern.go [hunk 2]
```diff
@@ -196,12 +197,33 @@ func readSpecialPattern(pb *patternBuild, valsIn []typedVal) (pathVals []typedVa
...
 	}
 	return
 }
 
+func readPrefixSpecial(pb *patternBuild, valsIn []typedVal) (pathVals []typedVal, err error) {
+	t, err := pb.jd.Token()
+	if err != nil {
+		return
+	}
+	pathVals = valsIn
```
Comment (embano1) [inline pattern.go:213]: What's the purpose of this assignment? This is not a (deep) copy.

File: README.md (hunk 1 above)
Comment (timbray) [reply README.md:120]: Good catch!

File: pattern.go [hunk 3]
```diff
@@ -196,12 +197,33 @@ func readSpecialPattern(pb *patternBuild, valsIn []typedVal) (pathVals []typedVa
...
+
+	prefixString, ok := t.(string)
+	if !ok {
+		err = errors.New("value for 'prefix' must be a string")
+		return
+	}
+	pathVals = append(pathVals, typedVal{vType: prefixType, val: `"` + prefixString + `"`})
```
Comment (embano1) [inline pattern.go:220]: nit: improve readability by assigning `typedVal{vType: prefixType, val: `"` + prefixString + `"`}` a variable and using the var in `append`

File: pattern.go [hunk 4]
```diff
@@ -196,12 +197,33 @@ func readSpecialPattern(pb *patternBuild, valsIn []typedVal) (pathVals []typedVa
...
+
+	// has to be } or tokenizer will throw error
+	_, err = pb.jd.Token()
+	return
```
Comment (embano1) [inline pattern.go:224]: In general I don't like named returns in Go functions (readability). Any reason we need named returns here?

File: value_matcher.go [hunk 5]
```diff
@@ -116,8 +117,8 @@ func (m *valueMatcher) addTransition(val typedVal) *fieldMatcher {
 			var newNfa *smallTable[*nfaStepList]
 			newNfa, nextField = makeShellStyleAutomaton(valBytes, nil)
 			newDfa = nfa2Dfa(newNfa)
-		default:
```
Comment (embano1) [inline value_matcher.go:119]: Why did you remove all the `default`s?

File: value_matcher.go [hunk 6]
```diff
@@ -183,6 +187,28 @@ func (m *valueMatcher) addTransition(val typedVal) *fieldMatcher {
...
 }
 
+func makePrefixAutomaton(val []byte, useThisTransition *fieldMatcher) (*smallTable[*dfaStep], *fieldMatcher) {
+	var nextField *fieldMatcher
+	if useThisTransition != nil {
+		nextField = useThisTransition
+	} else {
+		nextField = newFieldMatcher()
+	}
+	return onePrefixStep(val, 0, nextField), nextField
```
Comment (embano1) [inline value_matcher.go:197]: Nit: can be simplified (avoid `if/else` if possible)

```suggestion
	nextField := newFieldMatcher()
	if useThisTransition != nil {
		nextField = useThisTransition
	}
	return onePrefixStep(val, 0, nextField), nextField
```

File: value_matcher.go [hunk 7]
```diff
@@ -183,6 +187,28 @@ func (m *valueMatcher) addTransition(val typedVal) *fieldMatcher {
...
+}
+
+func onePrefixStep(val []byte, index int, nextField *fieldMatcher) *smallTable[*dfaStep] {
+	var nextStep *dfaStep
+
+	// have to stop one short to skip the closing "
+	if index == len(val)-2 {
+		nextStep = &dfaStep{table: newSmallTable[*dfaStep](), fieldTransitions: []*fieldMatcher{nextField}}
+	} else {
```
Comment (embano1) [inline value_matcher.go:206]: Nit: simplify, same as above

File: pattern.go (hunk 2 above)
Comment (timbray) [reply pattern.go:213]: Naked returns at work. More on this below

File: pattern.go (hunk 4 above)
Comment (timbray) [reply pattern.go:224]: I confess that this is the one area where I disagree with the Go community; as a result, there are named returns (heh, "naked" returns) here and there around Quamina. The linter doesn't seem to care.

For the moment, I don't feel any urgency about chasing them out. If someone cared enough I'd approve a PR that takes them all out systematically.

File: value_matcher.go (hunk 6 above)
Comment (timbray) [reply value_matcher.go:197]: But wouldn't that cause a wasted call to `newFieldMatcher` in the (common) `!= nil` case?

File: value_matcher.go (hunk 7 above)
Comment (timbray) [reply value_matcher.go:206]: Same response; wouldn't that cause a wasted memory allocation?

File: pattern.go (hunk 2 above)
Comment (embano1) [reply pattern.go:213]: Ok, besides the named (naked) returns, you don't care that this is a shall copy of `valsIn` here, i.e. immutability for the source slice is not required, correct?

File: pattern.go (hunk 3 above)
Comment (embano1) [reply pattern.go:220]: This is what I meant and I feel it improves readability, but this is nit:

```go
val := typedVal{
	vType: prefixType,
	val:   fmt.Sprintf("%q", prefixString), // or keep the concat if you're concerned about allocs
}
pathVals = append(pathVals, val)
```

File: value_matcher.go [hunk 8]
```diff
@@ -71,6 +71,7 @@ func (m *valueMatcher) transitionOn(val []byte) []*fieldMatcher {
 
 	default:
 		// no dfa, no singleton, nothing to do
+		// this probably can't happen because a flattener shouldn't preserve a field that hasn't appeared in a pattern
```
Comment (embano1) [inline value_matcher.go:74]: nit: comment line wrapping seems off

File: value_matcher.go (hunk 5 above)
Comment (embano1) [reply value_matcher.go:119]: You could test it IMHO (if you're concerned about coverage): create a new `valType` in your test and pass it. It should trigger the panic (which you can recover from and assert). Reason I like `panic` in those `switch` statements is that with the code base growing, slipping in a new `valType` and forgetting the places where this is `switch`-ed on is a possibility. Without a `default` case your code would happily continue and likely produce a bug.

File: value_matcher.go (hunk 6 above)
Comment (embano1) [reply value_matcher.go:197]: Was wondering if the compiler would be smart enough to figure out that branch and avoid the heap alloc as well. Looks like we don't have benchmarks here so can't judge. Is this a hot path or more like a one-off at startup?

File: value_matcher.go (hunk 5 above)
Comment (embano1) [reply value_matcher.go:119]: Update: here's an example how to start testing these cases

```go
func TestAddInvalidTransition(t *testing.T) {
	defer func() {
		if recover() == nil {
			t.Errorf("TestAddInvalidTransition should have panicked")
		}
	}()

	panicType := valType(999)

	m := newValueMatcher()
	v1 := typedVal{
		vType: panicType,
		val:   "one",
	}
	m.addTransition(v1)
}
```

Note: you could add more panic scenarios to cover the `if fields.startDfa != nil` branch. `default` in `singletonAutomaton` might not be needed as caught by earlier panic.

**Comment by embano1** [comment]
Just nits left, I'll approve and let you decide whether you want to make small adjustments based on the review. At least reverting the `default` scenario seems like a zero-waste but good engineering practice.

**Review by embano1 — APPROVED** [review]

File: value_matcher.go (hunk 5 above)
Comment (timbray) [reply value_matcher.go:119]: Excellent, thanks, will do.

File: value_matcher.go (hunk 6 above)
Comment (timbray) [reply value_matcher.go:197]: Actually, it turn out that the useThisTransition has become a one-off special case and in fact the argument is going to be removed from most of the build*Automaton calls. So I'm going to leave this as is for now and address in a future kaizen PR.

File: value_matcher.go (hunk 7 above)
Comment (timbray) [reply value_matcher.go:206]: as above

File: pattern.go (hunk 3 above)
Comment (timbray) [reply pattern.go:220]: OK, you're right, it adds readability.

File: value_matcher.go (hunk 5 above)
Comment (timbray) [reply value_matcher.go:119]: This definitely improved both the code and the coverage, thanks.

Comment (embano1) [reply value_matcher.go:119]: Awesome, glad to hear!

**Review by embano1 — APPROVED** [review]
LGTM
